packages/flutter/lib/src/painting/basic_types.dart

```diff
@@ -4,49 +4,9 @@
 
 import 'dart:ui' show TextDirection;
 
-export 'dart:ui' show
-  BlendMode,
-  BlurStyle,
-  Canvas,
-  Clip,
-  Color,
-  ColorFilter,
-  FilterQuality,
-  FontStyle,
-  FontWeight,
-  ImageShader,
-  Locale,
-  MaskFilter,
-  Offset,
-  Paint,
-  PaintingStyle,
-  Path,
-  PathFillType,
-  PathOperation,
-  Radius,
-  RRect,
-  RSTransform,
-  Rect,
-  Shader,
-  Size,
-  StrokeCap,
-  StrokeJoin,
-  TextAffinity,
-  TextAlign,
-  TextBaseline,
-  TextBox,
-  TextDecoration,
-  TextDecorationStyle,
-  TextDirection,
-  TextPosition,
-  TileMode,
-  VertexMode,
-  // TODO(werainkhatri): remove these after their deprecation period in engine
-  // https://github.com/flutter/flutter/pull/99505
-  hashValues, // ignore: deprecated_member_use
-  hashList; // ignore: deprecated_member_use
+export 'dart:ui' show TextDirection;
 
-export 'package:flutter/foundation.dart' show VoidCallback;
+export '_basic_types_extra_exports.dart';
```

Comment:
Why do we need this extra level of indirection?

Reply:
Basically the lint (unpublished for now) checks (1) that every type used in the direct public API is exported and (2) that no other types are exported (to help cleaning when you remove some api parts). The (2) part has 2 exceptions:

- no diagnostic if a file only contains exports
- no diagnostic for export without _show_ combinator

Those exceptions are the only way I found to make (2) part working correctly.

This means that there will be a diagnostic for `export 'package:flutter/foundation.dart' show VoidCallback;` if `VoidCallback` isn't directly used in the current file. To prevent this situation, I put all those _external_ export in a file with only exports and I forwarded those exports with this current line.

Another solution could be to move this exports to `lib/painting.dart` and fix imports in `lib/src/painting/*.dart`.

Reply:
Fixing up the existing `export 'dart:ui'` in `lib/painting.dart` sounds like a better solution to me. It makes the intent of this a little clearer.